### ISSUE TITLE
Allow options before input when using the CLI

### ISSUE THREAD

**Issue by Arcanemagus** [stylelint/stylelint#1605]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

v6.9.0 introduced a regression where _all_ parameters _require_ a parameter. This breaks parameters such as `--quiet` and `--extract` which do not use a parameter. To work around this a fake one must be provided, although it is ignored.

> Which version of stylelint are you using?

`6.9.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI with `stylelint --quiet` or `stylelint --extract`

> What did you expect to happen?

The "boolean" options to not require a parameter

> What actually happened (e.g. what warnings or errors you are getting)?

`--quiet` and `--extract` options require a fake parameter to be input.

**Comment by evilebottnawi** [comment]
@jeddy3 fixed? maybe add changelog?

**Comment by jeddy3** [comment]
https://github.com/stylelint/stylelint/pull/1606/commits/29b08325104823306addb2911fb99a7687e5a2e9

**Comment by jeddy3** [comment]
I suggest we release `6.9.1` when this is resolved. Any objections?

**Comment by evilebottnawi** [comment]
@jeddy3 need include https://github.com/stylelint/stylelint/pull/1609 in `6.9.1`

**Comment by jeddy3** [comment]
@Arcanemagus In what context are you using the `stylelint --quiet` and `stylelint --extract` commands *without any files nor piping in stdin*? I have a [pending PR](https://github.com/stylelint/stylelint/pull/1606), but I'm now [a little unsure](https://github.com/stylelint/stylelint/pull/1606#issuecomment-231375650) whether it's correct and would appreciate your input.

**Comment by Arcanemagus** [comment]
Sorry, here's the full reproduction:
foo.css:

```css
select {
background-color: blue;
}
```

.stylelintrc:

```json
{
  "rules": {
    "indentation": [2, {
      "indentInsideParens": "once",
      "severity": "warning"
    }]
  }
}
```

And here is what I see:
```ShellSession
PS C:\Temp\github\linter-stylelint_GH224> node --version
v6.3.0
PS C:\Temp\github\linter-stylelint_GH224> .\node_modules\.bin\stylelint --version
6.9.0
PS C:\Temp\github\linter-stylelint_GH224> .\node_modules\.bin\stylelint .\foo.css

foo.css
 2:1  ‼  Expected indentation of 2 spaces   indentation

PS C:\Temp\github\linter-stylelint_GH224> .\node_modules\.bin\stylelint --quiet .\foo.css

  Modern CSS linter

  Usage
    stylelint [input] [options]

  Input
    Files(s) or glob(s).
    If an input argument is wrapped in quotation marks, it will be passed to node-glob
    for cross-platform glob support.
    `node_modules` and `bower_components` are always ignored.
    You can also pass no input and use stdin, instead.

  Options
    --config            Path to a specific configuration file (JSON, YAML, or CommonJS),
                        or the name of a module in `node_modules` that points to one.
                        If no `--config` argument is provided, stylelint will search for
                        configuration  files in the following places, in this order:
                          - a `stylelint` property in `package.json`
                          - a `.stylelintrc` file (with or without filename extension:
                            `.json`, `.yaml`, and `.js` are available)
                          - a `stylelint.config.js` file exporting a JS object
                        The search will begin in the working directory and move up the
                        directory tree until a configuration file is found.
    --version           Get the currently installed version of stylelint.
    --custom-formatter  Path to a JS file exporting a custom formatting function
    --stdin-filename    Specify a filename to assign stdin input
    -f, --formatter     Specify a formatter: "json", "string" or "verbose". Default is "string".
    -i, --ignore-path   Specify a to a file containing patterns describing files
                        to ignore. The path can be absolute or relative to `cwd`.
                        By default, stylelint looks for `.stylelintignore` in the
                        same directory as the configuration file.
    -q, --quiet         Only register warnings for rules with an "error"-level severity
                        (ignore "warning"-level)
    -s, --syntax        Specify a non-standard syntax that should be used to
                        parse source stylesheets. Options: "scss", "less", "sugarss"
    -e, --extract       Extract and lint CSS from style tags in HTML structures
PS C:\Temp\github\linter-stylelint_GH224> .\node_modules\.bin\stylelint --quiet asdf .\foo.css
PS C:\Temp\github\linter-stylelint_GH224>
```

The steps are exactly the same for `--extract`.

**Comment by Arcanemagus** [comment]
Hmmm, I _just_ realized that `stylelint` is backwards from the CLI utilities I'm used to... the input is supposed to come first and options afterwards. I think the only other thing I've seen use that format is `find` 😕.

If you do the "correct" way, it works fine:

```ShellSession
PS C:\Temp\github\linter-stylelint_GH224> .\node_modules\.bin\stylelint .\foo.css --quiet
PS C:\Temp\github\linter-stylelint_GH224>
```

**Comment by evilebottnawi** [comment]
@Arcanemagus right, first `input`, then `option`
@jeddy3 maybe investigate how allow first `input` and last `input`?

**Comment by Arcanemagus** [comment]
@evilebottnawi `stylelint` already does (and always has) worked that way, it's just that v6.9.0 broke it for boolean flags.

**Comment by jeddy3** [comment]
> @jeddy3 maybe investigate how allow first input and last input?

My [pending PR](https://github.com/stylelint/stylelint/pull/1606) re-enables support for `options` first, `input` last. I tested it locally with `./node_modules/.bin/stylelint --quiet sources/css/temp.css` and it was fine, whereas `master` would display the help.

**Comment by jeddy3** [comment]
>  I tested it locally with ./node_modules/.bin/stylelint --quiet sources/css/temp.css and it was fine, whereas master would display the help.

Sorry, actually that's not true.  I misread the output. Help isn't display but nor are the files correctly set. Instead it's trying to read from stdin.

@Arcanemagus Has `options` first `input` last ever worked?

**Comment by Arcanemagus** [comment]
Like I said, it works in v6.8.0, and also in v6.9.0 but you need a fake parameter for that version:
![image](https://cloud.githubusercontent.com/assets/427137/16695814/9e730d18-44f6-11e6-9f6b-78aae837b6e1.png)

**Comment by Arcanemagus** [comment]
Hmm, that above example doesn't show that it's actually working, since it specifically has no output, this however shows it _is_ grabbing the file:
![image](https://cloud.githubusercontent.com/assets/427137/16695991/9d310df0-44f7-11e6-97b5-94409f8b56e1.png)

**Comment by jeddy3** [comment]
I'm afraid I can't seem to replicate this in `6.8.0` either i.e. `stylelint --formatter verbose ./foo.css` works (I guess because it's not a boolean flag), but `stylelint --quiet ./foo.css` seems to look to stdin. I assume that's because `./foo.css` is considered a parameter of the `--quiet` flag.

Irregardless of whether it was supported it in the past, we can certainly investigate the feasibility of adding it going forward. We use [meow](https://github.com/sindresorhus/meow) for the CLI, so I guess it all depends on what it supports. I'll delete my [flawed PR](https://github.com/stylelint/stylelint/pull/1606) and relabel this as "help wanted" and "needs investigation".

**Comment by davidtheclark** [comment]
> I assume that's because ./foo.css is considered a parameter of the --quiet flag

I'd guess that's the explanation here, and it makes sense that this would happen.

@Arcanemagus From looking around at various CLI tools I see a mix in the documentation, regarding whether options come first or second. Not sure if these tools have logic to allow for *both* orders or not.

**Comment by Arcanemagus** [comment]
Just a note, the [CSS processors](http://stylelint.io/user-guide/css-processors/) documentation page shows an example with the options coming first 😛.

**Comment by jeddy3** [comment]
Good catch. Shall we correct the docs and then close this issue, or do you want to further investigate the feasibility of supporting flags before input?

**Comment by Arcanemagus** [comment]
Personally I would prefer the ability to have option flags before input, but I don't have time to dedicate to looking into it here myself 😞.

**Comment by jeddy3** [comment]
@Arcanemagus I'll fix the docs and close the issue for now. We can reopen it if anyone else requests this feature or if you have time to contribute it yourself.
